# Lab 8: Naive Bayes Classifier

---



## Importing the libraries

In [1]:
from sklearn.datasets import fetch_openml
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

## Loading the dataset

In [2]:
mnist = fetch_openml('mnist_784')
X, y = mnist['data'], mnist['target']

c:\Users\jamal\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\datasets\_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


## Data preprocessing

In [3]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Columns: 784 entries, pixel1 to pixel784
dtypes: float64(784)
memory usage: 418.7 MB


In [4]:
X.describe()

pixel1   pixel2   pixel3   pixel4   pixel5   pixel6   pixel7   pixel8  \
count  70000.0  70000.0  70000.0  70000.0  70000.0  70000.0  70000.0  70000.0   
mean       0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
std        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
min        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
25%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
50%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
75%        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   
max        0.0      0.0      0.0      0.0      0.0      0.0      0.0      0.0   

        pixel9  pixel10  ...      pixel775      pixel776      pixel777  \
count  70000.0  70000.0  ...  70000.000000  70000.000000  70000.000000   
mean       0.0      0.0  ...      0.197414      0.099543      0.046629   
std        0.0      0.0  ...      5.991206      4.256304      2.783732   
min        0.0      0.0  ...      0.000000      0.000000      0.000000   
25%        0.0      0.0  ...      0.000000      0.000000      0.000000   
50%        0.0      0.0  ...      0.000000      0.000000      0.000000   
75%        0.0      0.0  ...      0.000000      0.000000      0.000000   
max        0.0      0.0  ...    254.000000    254.000000    253.000000   

           pixel778      pixel779      pixel780  pixel781  pixel782  pixel783  \
count  70000.000000  70000.000000  70000.000000   70000.0   70000.0   70000.0   
mean       0.016614      0.012957      0.001714       0.0       0.0       0.0   
std        1.561822      1.553796      0.320889       0.0       0.0       0.0   
min        0.000000      0.000000      0.000000       0.0       0.0       0.0   
25%        0.000000      0.000000      0.000000       0.0       0.0       0.0   
50%        0.000000      0.000000      0.000000       0.0       0.0       0.0   
75%        0.000000      0.000000      0.000000       0.0       0.0       0.0   
max      253.000000    254.000000     62.000000       0.0       0.0       0.0   

       pixel784  
count   70000.0  
mean        0.0  
std         0.0  
min         0.0  
25%         0.0  
50%         0.0  
75%         0.0  
max         0.0  

[8 rows x 784 columns]

In [5]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 70000 entries, 0 to 69999
Series name: class
Non-Null Count  Dtype   
--------------  -----   
70000 non-null  category
dtypes: category(1)
memory usage: 68.9 KB


In [6]:
y.describe()

count     70000
unique       10
top           1
freq       7877
Name: class, dtype: object

## Splitting the dataset into Training set and Test set

In [7]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

In [8]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 56000 entries, 47339 to 15795
Columns: 784 entries, pixel1 to pixel784
dtypes: float64(784)
memory usage: 335.4 MB


In [9]:
y_train.info()

<class 'pandas.core.series.Series'>
Index: 56000 entries, 47339 to 15795
Series name: class
Non-Null Count  Dtype   
--------------  -----   
56000 non-null  category
dtypes: category(1)
memory usage: 492.6 KB


In [10]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Index: 14000 entries, 46730 to 63452
Columns: 784 entries, pixel1 to pixel784
dtypes: float64(784)
memory usage: 83.8 MB


In [11]:
y_test.info()

<class 'pandas.core.series.Series'>
Index: 14000 entries, 46730 to 63452
Series name: class
Non-Null Count  Dtype   
--------------  -----   
14000 non-null  category
dtypes: category(1)
memory usage: 123.4 KB


## Applying feature scaling

In [12]:
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Training Gaussian Naïve Bayes classifier

In [13]:
accuracies = []

In [14]:
classifier_gnb = GaussianNB()
classifier_gnb.fit(X_train, y_train)
predictions_gnb = classifier_gnb.predict(X_test)
accuracy_gnb = accuracy_score(y_test, predictions_gnb)
accuracies.append(accuracy_gnb)

## Training Decision Tree classifier

In [15]:
classifier_dt_ = DecisionTreeClassifier(criterion = 'entropy', random_state=42)
classifier_dt_.fit(X_train, y_train)
predictions_dt_ = classifier_dt_.predict(X_test)
accuracy_dt_ = accuracy_score(y_test, predictions_dt_)
accuracies.append(accuracy_dt_)

## Training Support Vector classifier

In [16]:
classifier_svc_linear = SVC(kernel = 'linear', random_state=42)
classifier_svc_linear.fit(X_train, y_train)
predictions_svc_linear = classifier_svc_linear.predict(X_test)
accuracy_svc_linear = accuracy_score(y_test, predictions_svc_linear)
accuracies.append(accuracy_svc_linear)

In [17]:
classifier_svc_rbf = SVC(kernel = 'rbf', random_state=42)
classifier_svc_rbf.fit(X_train, y_train)
predictions_svc_rbf = classifier_svc_rbf.predict(X_test)
accuracy_svc_rbf = accuracy_score(y_test, predictions_svc_rbf)
accuracies.append(accuracy_svc_rbf)

## Training Logistic Regression classifier

In [18]:
classifier_lr = LogisticRegression()
classifier_lr.fit(X_train, y_train)
predictions_lr = classifier_lr.predict(X_test)
accuracy_lr = accuracy_score(y_test, predictions_lr)
accuracies.append(accuracy_lr)

c:\Users\jamal\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


## Evaluating the models

In [19]:
classifiers = ['Gaussian Naïve Bayes', 'Decision Tree', 'Support Vector (Linear) Classifier', 'Support Vector (RBF) Classifier', 'Logistic Regression']

for i in range(5):
    print(f"Classifier - {classifiers[i]}")
    print(f"Accuracy: {accuracies[i]*100:.2f} %\n")

Classifier - Gaussian Naïve Bayes
Accuracy: 52.50 %

Classifier - Decision Tree
Accuracy: 88.02 %

Classifier - Support Vector (Linear) Classifier
Accuracy: 92.11 %

Classifier - Support Vector (RBF) Classifier
Accuracy: 96.30 %

Classifier - Logistic Regression
Accuracy: 91.50 %



## Conclusion

* Among the tested models, the **Support Vector Classifier (SVC) with the Radial Basis Function (RBF) kernel** performed the best, achieving the **highest accuracy of 96.30%**. This superiority can be attributed to the RBF kernel's ability to capture complex nonlinear relationships in the high-dimensional MNIST dataset.

* While the Decision Tree and Logistic Regression models showed good performance, they were outperformed by the SVMs on this specific dataset. Decision Trees can capture complex relationships but are prone to overfitting, while Logistic Regression performs well for linearly separable data but may struggle with more complex patterns.

* Gaussian Naïve Bayes lagged behind the other models significantly, possibly due to its oversimplified assumption of feature independence, which doesn’t align well with the correlated pixel values in image data like MNIST.

In summary, for the MNIST dataset, the Support Vector Classifier with the Radial Basis Function (RBF) kernel stands out as the best-performing model, offering a balance between accuracy and generalization to unseen digit images.